runtime/common/src/claims.rs

```diff
@@ -336,6 +464,77 @@ impl<T: Trait> sp_runtime::traits::ValidateUnsigned for Module<T> {
 	}
 }
 
+/// Validate `attest` calls prior to execution. Needed to avoid a DoS attack since they are
+/// otherwise free to place on chain.
+#[derive(Encode, Decode, Clone, Eq, PartialEq)]
+pub struct PrevalidateAttests<T: Trait + Send + Sync>(sp_std::marker::PhantomData<T>) where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>;
+
+impl<T: Trait + Send + Sync> Debug for PrevalidateAttests<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "PrevalidateAttests")
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<T: Trait + Send + Sync> PrevalidateAttests<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	/// Create new `SignedExtension` to check runtime version.
+	pub fn new() -> Self {
+		Self(sp_std::marker::PhantomData)
+	}
+}
+
+impl<T: Trait + Send + Sync> SignedExtension for PrevalidateAttests<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	type AccountId = T::AccountId;
+	type Call = <T as system::Trait>::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+	const IDENTIFIER: &'static str = "PrevalidateAttests";
+
+	fn additional_signed(&self) -> Result<Self::AdditionalSigned, TransactionValidityError> {
+		Ok(())
+	}
+
+	fn validate(
+		&self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		if let Some(local_call) = call.is_sub_type() {
+			if let Call::attest(attested_statement) = local_call {
+				let signer = match Preclaims::<T>::get(who) {
+					Some(x) => x,
+					None =>
+						return Err(InvalidTransaction::Custom(
+							ValidityError::SignerHasNoClaim.into()
+						).into()),
+				};
+				if let Some(s) = Signing::get(signer) {
+					if &attested_statement[..] != s.to_text() {
```

Comment:
Can the exact statement ("Default"/"Alternative") already be tested on the UI? I tried to attest with some random bytes, and the tx still passed.

Here's my repro:
- modify chain spec to add a preclaimed account (AMAURY THIBAUT)
- run this branch
- see recording


![ezgif com-video-to-gif](https://user-images.githubusercontent.com/1293565/81914232-52ea1f00-95d1-11ea-9335-3e8389663316.gif)

Notice also how the upper notification bar stating that this account needs to attest disappears.

Reply:
maybe try on a more recent version. the text was originally being checked only with the SignedExtension, which I've only just enabled.

it definitely shouldn't work, and there's a unit test there for it...